feat: support retries without exponential backoff

DonutEspresso opened the pull request:
By default, restify clients retries up to 4 times on failure to establish connection. This PR adds:

* an options flag `exponentialBackoff: false` which can be set on the client or on per request basis. this enables immediate retries without any backoff. 
* a `req.attempts()` method which returns the number of attempts made for this specific request

hekike commented on lib/HttpClient.js:
```diff
@@ -667,28 +674,54 @@ HttpClient.prototype.request = function request(opts, cb) {
+    if (opts.exponentialBackoff === false) {
+        // if no backoff, we have to keep track of num retries ourselves instead
+        // of relying on backoff module.
+        rawRequest(opts, retryInfo, function retryRawRequest(err, req) {
```
How about returning the `rawRequest` and dropping the else "wrapper"?

hekike commented on lib/HttpClient.js:
```diff
@@ -667,28 +674,54 @@ HttpClient.prototype.request = function request(opts, cb) {
+        // if no backoff, we have to keep track of num retries ourselves instead
+        // of relying on backoff module.
+        rawRequest(opts, retryInfo, function retryRawRequest(err, req) {
+            if (err && retryInfo.attempts <= retry.retries) {
```
Don't you want to check for `statusCode`? We don't want to retry for client error like 404, right?

DonutEspresso replied:
This is already handled in `rawRequest`, but the behavior is that an err is only returned if it's a networking issue. Turning a statusCode into an error actually happens after this.

hekike commented on lib/HttpClient.js:
```diff
@@ -667,28 +674,54 @@ HttpClient.prototype.request = function request(opts, cb) {
+        call.failAfter(retry.retries);
+        // TODO: is this emitted event actually useful? if you have multiple
+        // inflight requests, how do you differentiate them?
+        call.on('backoff', this.emit.bind(this, 'attempt'));
```
Can we avoid using bind with `self`? Bind is slower.

DonutEspresso replied:
This is actually existing code, but I can turn this into a closure if you prefer.

DonutEspresso replied:
Touche, the event is from existing code, but I can move that into the same listener.

hekike commented on lib/HttpClient.js:
```diff
@@ -698,6 +731,8 @@ HttpClient.prototype._options = function (method, options) {
         cert: options.cert || self.cert,
         ciphers: options.ciphers || self.ciphers,
         connectTimeout: options.connectTimeout || self.connectTimeout,
+        exponentialBackoff: typeof options.exponentialBackoff !== 'undefined' ?
```
`options.exponentialBackoff !== undefined`?

DonutEspresso replied:
By default the value should be true, so the user sets it with `false` which means I can't do `false || true`. There's probably a more elegant way to do this but this is pretty explicit.

hekike replied:
Ahh okay, my bad. Avoid me

hekike commented on lib/HttpClient.js:
```diff
@@ -222,6 +222,9 @@ function rawRequest(opts, cb) {
         emitResult((err || null), req, res);
     });
     req.log = log;
+    req.attempts = function getAttempts() {
```
How about `req.getAttempts`, see discussion here: https://github.com/restify/clients/pull/139#issuecomment-363228304

DonutEspresso commented on lib/HttpClient.js:
```diff
@@ -111,7 +111,7 @@ function RequestTimeoutError(ms) {
 util.inherits(RequestTimeoutError, Error);
 
-function rawRequest(opts, cb) {
+function rawRequest(opts, retryInfo, cb) {
```
`retryInfo` is mutable (through the course of multiple retries), whereas `opts` isn't - was trying to keep the two separate.

DonutEspresso commented on lib/HttpClient.js:
```diff
@@ -667,28 +674,54 @@ HttpClient.prototype.request = function request(opts, cb) {
+    } else {
+        // if backoff is specified (default behavior) then use backoff module
+        // to keep track of retry count.
+        call = backoff.call(rawRequest, opts, retryInfo, cb);
```
Good call, har har

DonutEspresso commented:
Comments addressed, once more?

DonutEspresso commented:
Rebased against master.

@hekike retries aren't new and they're only meant to retry against network errors, but I've added tests to ensure it doesn't retry against http status codes. 

I also just added a breaking change to rename the emitted `attempt` event to `retry`. Since we now expose `req.getAttempts()` it is weird that that is not in sync with the number of times the `attempt` event was emitted. Thus, renaming it to `retry` to better capture the meaning of that event.

DonutEspresso commented:
I cleaned up the impl a little bit, so it's more symmetrical between exponential retry vs standard retry scenarios. Also added an `_onRetry` lifecycle hook that allows inherited clients an opportunity to do work before the next retry is fired.

hekike commented on lib/HttpClient.js:
```diff
@@ -674,31 +681,71 @@ HttpClient.prototype.request = function request(opts, cb) {
+
 
-    call.start();
+HttpClient.prototype._onRetry = function _onRetry(err, retryInfo, callback) {
```
missing JSDoc, but I see it's missing from the whole project :)

DonutEspresso commented:
Yep! Although, recent investigations into cueball might make this PR DOA - will keep it open for now but there's a high possibility this won't get merged in.

DonutEspresso commented:
Closing this in favor of redirecting folks to use [cueball](https://github.com/joyent/node-cueball/).
